android/src/com/mapswithme/maps/widget/placepage/PlaceOpeningHoursAdapter.java

```diff
@@ -8,55 +8,92 @@
 import androidx.annotation.NonNull;
 import androidx.recyclerview.widget.RecyclerView;
 import com.mapswithme.maps.R;
-import static com.mapswithme.maps.editor.data.TimeFormatUtils.formatWeekdays;
-import static com.mapswithme.maps.editor.data.TimeFormatUtils.formatNonBusinessTime;
-
 import com.mapswithme.maps.editor.data.Timespan;
 import com.mapswithme.maps.editor.data.Timetable;
 import com.mapswithme.util.UiUtils;
 
 import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 
+import static com.mapswithme.maps.editor.data.TimeFormatUtils.formatNonBusinessTime;
+import static com.mapswithme.maps.editor.data.TimeFormatUtils.formatWeekdaysRange;
+
 public class PlaceOpeningHoursAdapter extends RecyclerView.Adapter<PlaceOpeningHoursAdapter.ViewHolder>
 {
-  private Timetable[] mTimetables = {};
-  private int[] closedDays = null;
+  private List<WeekScheduleData> mWeekSchedule = Collections.emptyList();
 
   public PlaceOpeningHoursAdapter() {}
 
-  public PlaceOpeningHoursAdapter(Timetable[] timetables)
+  public PlaceOpeningHoursAdapter(Timetable[] timetables, int firstDayOfWeek)
   {
-    setTimetables(timetables);
+    setTimetables(timetables, firstDayOfWeek);
   }
 
-  public void setTimetables(Timetable[] timetables)
+  public void setTimetables(Timetable[] timetables, int firstDayOfWeek)
   {
-    mTimetables = timetables;
-    closedDays = findUnhandledDays(timetables);
+    int[] weekDays = buildWeekByFirstDay(firstDayOfWeek);
+
+    final List<WeekScheduleData> scheduleData = new ArrayList<>();
+
+    // Timetables array contains only working days. We need to fill non working gaps
+    for (int i=0; i < weekDays.length; i++)
+    {
+      int weekDay = weekDays[i];
+
+      Timetable tt = findScheduleForWeekDay(timetables, weekDay);
+      if (tt != null)
+      {
+        int startWeekDay = weekDays[i];
+        while (i < weekDays.length && tt.containsWeekday(weekDays[i]))
+          i++;
+
+        i--;
+        int endWeekDay = weekDays[i];
+        scheduleData.add(new WeekScheduleData(startWeekDay, endWeekDay, tt));
+      }
+      else
+      {
+        int startWeekDay = weekDays[i];
+        // Search next working day in timetables
+        while (i+1 < weekDays.length)
+        {
+          if (findScheduleForWeekDay(timetables, weekDays[i+1]) != null)
+            break;
+          i++;
+        }
+
+        scheduleData.add(new WeekScheduleData(startWeekDay, weekDays[i], null));
+      }
+    }
+
+    mWeekSchedule = scheduleData;
+
     notifyDataSetChanged();
   }
 
-  private int[] findUnhandledDays(Timetable[] timetables)
+  public static int[] buildWeekByFirstDay(int firstDayOfWeek)
```

Comment:
A possible easier implementation:
```java
      final List<Integer> list = Arrays.asList(1, 2, 3, 4, 5, 6, 7);  // Or Calendar.SUNDAY, Calendar.MONDAY, … 
      Collections.rotate(list, 1 - firstDayOfWeek);
      return Ints.toArray(list);  // Or use List<Integer> in the code.
```

Reply:
Implemented using `Collections.rotate` method. Replaced week days array with list for simplicity.